crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -526,48 +575,73 @@ where
             state_root_message_sender,
             multi_added_removed_keys,
         } = multiproof_input;
+
+        let account_targets = proof_targets.len();
+        let storage_targets = proof_targets.values().map(|slots| slots.len()).sum::<usize>();
+
+        trace!(
+            target: "engine::root",
+            proof_sequence_number,
+            ?proof_targets,
+            account_targets,
+            storage_targets,
+            ?source,
+            "Starting multiproof calculation",
+        );
+
+        let account_proof_task_handle = self.account_proof_task_handle.clone();
         let storage_proof_task_handle = self.storage_proof_task_handle.clone();
-        let missed_leaves_storage_roots = self.missed_leaves_storage_roots.clone();
+        // Clone and freeze prefix sets to get the current accumulated state from all transactions
+        let frozen_prefix_sets = (*config.prefix_sets).clone().freeze();
 
-        self.executor.spawn_blocking(move || {
-            let account_targets = proof_targets.len();
-            let storage_targets = proof_targets.values().map(|slots| slots.len()).sum::<usize>();
+        debug!(
+            target: "engine::root",
+            proof_sequence_number,
+            ?source,
+            "spawn_multiproof: FREEZING prefix_sets account_len={} storage_len={} for worker",
+            frozen_prefix_sets.account_prefix_set.len(),
+            frozen_prefix_sets.storage_prefix_sets.len(),
+        );
 
-            trace!(
-                target: "engine::root",
-                proof_sequence_number,
-                ?proof_targets,
-                account_targets,
-                storage_targets,
-                ?source,
-                "Starting multiproof calculation",
+        self.executor.spawn_blocking(move || {
```

Comment:
I think after this round of changes is done we should take another pass and see if we can factor out this spawn_blocking.